nemo/collections/asr/parts/submodules/adapters/transformer_multi_head_attention_adapter_module.py

```diff
@@ -0,0 +1,128 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+from dataclasses import dataclass, field
+from typing import Any, Optional
+
+import torch
+from torch import nn as nn
+
+from nemo.collections.asr.modules.transformer import transformer_modules
+from nemo.collections.asr.parts.submodules.adapters.multi_head_attention_adapter_module import (
+    MHAResidualAddAdapterStrategy,
+    MHAResidualAddAdapterStrategyConfig,
+)
+from nemo.collections.common.parts import adapter_modules
+from nemo.core.classes.mixins import adapter_mixin_strategies, adapter_mixins
+
+
+class TransformerMultiHeadAttentionAdapter(transformer_modules.MultiHeadAttention, adapter_modules.AdapterModuleUtil):
```

Comment:
why do we need specific MHA for adapter, can;t we use existing MHA layer and add it as adapter?

Reply:
The issue is that this module performs a pre-layer norm before calling the attention code that it subclasses. It seems this is needed for stable training, otherwise the output of previous attention layer is so large that this layer explodes.